src/journal-remote/journal-gatewayd.c

```diff
@@ -637,6 +751,63 @@ static int request_handler_fields(
         return MHD_queue_response(connection, MHD_HTTP_OK, response);
 }
 
+static int output_field_name(FILE *f, OutputMode m, const char *d) {
+        if (m == OUTPUT_JSON)
+                return fprintf(f, "{ \"field\" : \"%s\" }\n", d);
+        else
+                return fprintf(f, "%s\n", d);
+}
+
+static int request_handler_field_names(
+                struct MHD_Connection *connection,
+                void *connection_cls) {
+
+        _cleanup_(MHD_destroy_responsep) struct MHD_Response *response = NULL;
+        RequestMeta *m = connection_cls;
+        int fd = -1;
+        FILE *tmp = NULL;
```

Comment:
Hmm, actually my previous suggestion is not correct, but these should be closed on failure...

Reply:
Right ! But I'm not sure it's the right way to code it. Other options :

- On branch ["copypaste"](https://github.com/Samuel-BF/systemd/blob/journal-reader-copypaste/src/journal-remote/journal-gatewayd.c#L761) I implemented `field_names` requests like for `fields_request` : [request_handler_field_names](https://github.com/Samuel-BF/systemd/blob/journal-reader-copypaste/src/journal-remote/journal-gatewayd.c#L834) calls [MHD_create_response_from_callback](https://github.com/Samuel-BF/systemd/blob/31a52c03253f346af5156d6c179cfe9ff985f2df/src/journal-remote/journal-gatewayd.c#L854) with a (quite convoluted) function that reads the next field name and puts into MHD buffer.

This way, you have 3 functions ([request_handler_entries](https://github.com/Samuel-BF/systemd/blob/journal-reader-copypaste/src/journal-remote/journal-gatewayd.c#L566), [request_handler_fields](https://github.com/Samuel-BF/systemd/blob/journal-reader-copypaste/src/journal-remote/journal-gatewayd.c#L721) and [request_handler_field_names](https://github.com/Samuel-BF/systemd/blob/31a52c03253f346af5156d6c179cfe9ff985f2df/src/journal-remote/journal-gatewayd.c#L834)) which are very similar. Is it better ? (same logic everywhere)

- We could merge the 3 functions by adding members to RequestMeta objects. That would be shorter, would avoid copypaste, but would also add another level of complexity. Do you think it remains manageable if we do this way ? (I'm OK to (try to) code it this way if you think it's worth trying !)

And, btw : thanks a lot for your review and the time/energy taken for this.

Reply:
Fixed by adding `TAKE_PTR(tmp)` after `TAKE_FD(fd)`.

If I'm correct, in the everything-works case : 

-  MHD_create_response_from_fd will [close fd upon completion](https://www.gnu.org/software/libmicrohttpd/manual/html_node/microhttpd_002dresponse-create.html#index-MHD_005fcreate_005fresponse_005ffrom_005ffd)
- this will free (close ?) the `tmp` file stream as a consequence (man 2 close : "the resources associated with the open file description are freed")

And if there's an error somewhere:

- compiler inserted `close` and `fclose` calls everywhere (thanks to `_cleanup_f?close_`). It is redundant (we could remove either `_cleanup_close_`, either `_cleanup_fclose_`), as closing the fd will free the FILE * stream and the other way round, but ... it's more clean anyway.

Am I right ?

I also coded [the refactoring](https://github.com/Samuel-BF/systemd/blob/journal-reader-refactor/src/journal-remote/journal-gatewayd.c), using init/iterator functions in RequestMeta for `/fields`, `/fields/xxx` and `/entries` endpoints. The code is shorter than in the copypaste version, but harder to follow and understand. So the 3 options are : 
- this one (here), where we use a different logic for field and field names (a bit inconsistent)
- [copy-paste](https://github.com/Samuel-BF/systemd/blob/journal-reader-copypaste/src/journal-remote/journal-gatewayd.c#L761) which uses the very same logic for both, but is a bit redundant.
- [refactor](https://github.com/Samuel-BF/systemd/blob/journal-reader-refactor/src/journal-remote/journal-gatewayd.c) which is shorter, but less readable (and imho a bit more messy)